enhancements/network/baremetal-ipi-network-configuration.md

```diff
@@ -0,0 +1,241 @@
+available the config files can be directly inserted into the new interface.
+However, this is not a solution because it still requires manually copying
+configuration to multiple locations which introduces the potential for errors.
+
+## Design Details
+
+In the IPI flow day-1 network configuration is required in 2 different cases:
+
+* Deployment of the controlplane hosts via terraform, using input provided to openshift-install
+* Deployment of compute/worker hosts (during initial deployment and scale-out), via Machine API providers for each platform
+
+In the sections below we will describe the user-facing API that contains network configuration, and the proposed integration for each of these cases.
+
+### User-facing API
+
+RHCOS already provides a mechanism to specify NetworkManager keyfiles during deployment of a new node. We need to expose that functionality during the IPI install process. There are a couple of options on how to do that:
+
+* A new section in install-config.
+* A secret that contains base64-encoded content for the keyfiles.
+
+These are not mutually exclusive. If we implement the install-config option, we will still need to persist the configuration in a secret so it can be used for day-2. In fact, the initial implementation should likely be based on secrets, and then install-config integration can be added on top of that to improve the user experience. This way the install-config work won't block implementation of the feature.
+
+The data provided by the user will need to have the following structure:
+```yaml
+[hostname]:
+  [interface name]:
+    [contents of NetworkManager key file for interface]
+  [second interface name]:
+    [contents of NetworkManager key file for second interface]
+  etc...
+```
+
+For example:
+```yaml
+openshift-master-0:
+  eth0:
+    [connection]
+    id=eth0
+    uuid=18e0cec7-041c-4fb3-957c-a60c80dd9b85
+    type=ethernet
+    etc...
+openshift-master-1:
+  eth0:
+    [connection]
+    id=eth0
+    etc...
+```
+
+In install-config this would look like:
+```yaml
+platform:
+  baremetal:
+    hosts:
+      - name: openshift-master-0
+        networkConfig:
+          eth0:
+            [connection]
+            id=eth0
+            uuid=18e0cec7-041c-4fb3-957c-a60c80dd9b85
+            type=ethernet
+            etc...
+```
+
+Because the initial implementation will be baremetal-specific, we can put the
+network configuration data into the baremetal host field, which will allow easy
+mapping to the machine in question.
+
+#### API Open Questions
+
+* What is the best way to encode the data above into secret(s)?
+
+### Processing user configuration
+
+#### Deployment of the controlplane hosts via terraform
+
+We will map the keyfiles to their appropriate Machine using the host field
+of the baremetal install-config. The keyfiles will then be added to custom
+images for each Machine built by Terraform and Ironic.
+
+TODO - Does anyone have more detail to add here?
```

Comment:
@kirankt and/or @dtantsur perhaps you can help with more details here?

I think it will be something like:
* During startup of the provisioning services on the bootstrap VM, we will build a customized image (for each master?) containing the network configuration specified in the install-config
* This image will be passed to Terraform/Ironic to ensure each host is deployed with the appropriate configuration
* The configuration will be persisted to the cluster via a secret, so that day-2 rebuild of the controlplane hosts is possible

Re the secret, I'm not entirely clear if we'll expect to create a secret per-host, or a single secret with the list of hosts?  That probably depends on the image-building implementation so I guess either could work.

Reply:
> Re the secret, I'm not entirely clear if we'll expect to create a secret per-host, or a single secret with the list of hosts? That probably depends on the image-building implementation so I guess either could work.

Yeah, that's why I left that as an open question. I don't have enough familiarity to say which works better. I'm hoping we can get input from the people closer to the implementation.

Reply:
In the current implementation, we only deal with one image for all control plane hosts. We could make this separate, which would mean three different copies of, essentially the same, OS image. I like the idea of separation TBH, because it would open up the possibility of deploying non-homogeneous control plane hardware in the future. If we go this route, then the installer will create 3 secrets per control plane host.

Reply:
Okay, that works for me. One secret per host is conceptually easier too. If we decide we want hosts to be able to share secrets at some point that can probably be added.

Reply:
Yeah, a secret per host is easier to reason about.